### PR TITLE
Allows nimbus to accept context defined in the consumer

### PR DESCRIPTION
**Pull request by tarikeshaq**
I'll need to add more thorough testing, but this is an initial stab at the problem to see if we wanna go through this route.

fixes https://mozilla-hub.atlassian.net/browse/SDK-345

A couple of thoughts about this, and I'm not sure if this warrants an ADR...

Pros:
  - Consumers can now define their own `targeting` fields, and as long as those align with what the settings server sends us, then the nimbus SDK will be able to evaluate it correctly, without the SDK needing to know about the specific fields. This saves us developer time of having to go on and add more fields to the `AppContext` struct.
  - As non-browser consumers (or consumers who simply don't have some of the fields other consumers have, I'm looking at you `android_sdk_version`) it wouldn't make sense for the SDK to include those fields. 

Cons:
  - Biggest con I can think of is that there's less type checking. The consumers would have to be extra careful to get their fields right, when right now the uniffi-generated struct make it impossible to get wrong.
  - Minor con is that if we want to access the `AppContext.specific_context` fields for non-targeting-related reasons, we'll probably need to be a bit more careful on the Rust side to add error handling

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/nimbus/src/matcher.rs
Comment (tarikeshaq) [inline components/nimbus/src/matcher.rs:73]: Just to be 1000% sure, the fact that this is optional, it doesn't break backward compatibility right?

File: components/nimbus/src/nimbus.udl
Comment (tarikeshaq) [inline components/nimbus/src/nimbus.udl:16]: `DOMString` here is a awkward, but I think that's a uniffi limitation? I'll double-check, saw it in the docs

Comment (tarikeshaq) [reply components/nimbus/src/nimbus.udl:16]: Not too sure about the name `specific_context`, happy to take suggestions for something more representative

File: components/nimbus/src/matcher.rs
Comment (travis79) [reply components/nimbus/src/matcher.rs:73]: Yeah, as long as this is optional, then it shouldn't fail serializing/deserializing if it isn't there.

File: components/nimbus/src/nimbus.udl
Comment (travis79) [reply components/nimbus/src/nimbus.udl:16]: I don't have any issues with `specific_context`, maybe adding `app_` to it might make it more clear?  So `app_specific_context` or even `app_defined_context`?  Again, not blocking on this (naming things is hard).

File: components/nimbus/src/matcher.rs
Comment (jhugman) [inline components/nimbus/src/matcher.rs:73]: This is a really elegant way of solving this!

File: components/nimbus/ios/Nimbus/NimbusApi.swift
Comment (jhugman) [inline components/nimbus/ios/Nimbus/NimbusApi.swift:217]: I'm not wild about the naming, although I prefer `appSpecificContext` to `specificContext`. FWIW [`Context` has a very specific meaning][1] in Android development.

I prefer `customAttributes` or `appSpecificTargetingAttributes`, or some such. Everyone likes custom things.

[1]: https://developer.android.com/reference/android/content/Context

I appreciate you've already renamed once, so please take this as optional.

File: components/nimbus/src/matcher.rs
Comment (jhugman) [inline components/nimbus/src/matcher.rs:56]:
```suggestion
/// - `app_specific_context`: Context contains attributes specific to the application, derived by the application.
```

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:270]: We might also simplify a lot of code here by making all this non-optional.

```suggestion
    val appSpecificContext: Map<String, String> = mapOf()
```

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:268]: By this point, this is app developer focused documentation.

I suggest the docs should be written for them: e.g. application derived attributes measured by the application, but useful for targeting of experiments. e.g. `mapOf("userType" to "casual", "isFirstTime": "true")`.

File: components/nimbus/src/evaluator.rs
Comment (jhugman) [inline components/nimbus/src/evaluator.rs:329]: I had only just noticed that all the existing attributes were `String`. Question: does JEXL allows us to get numbers from strings? If so, this should definitely be documented/tested here.

**Review by jhugman — APPROVED** [review]
This is lovely. A few minor minor things, and a couple of questions, but nothing to hold back with another review cycle. 

I just noticed this is a Draft, so please ask as many questions as you want before asking for another review.

File: components/nimbus/src/evaluator.rs
Comment (travis79) [reply components/nimbus/src/evaluator.rs:329]: I'm pretty sure it does, since we can use comparison operators..   

See:
https://github.com/mozilla/jexl-rs/blob/351aaeb6eb172b2538a46a6aa52cc14bd1b0d909/jexl-parser/src/ast.rs#L47

and looking at some of the tests in jexl-rs

File: components/nimbus/src/matcher.rs
Comment (travis79) [inline components/nimbus/src/matcher.rs:73]: Best name yet! ❤️

**Review by travis79 — APPROVED** [review]
This is awesome!  Thanks @tarikeshaq

File: components/nimbus/src/evaluator.rs
Comment (tarikeshaq) [reply components/nimbus/src/evaluator.rs:329]: I'll create a follow-up ticket in the testing epic to add more comprehensive jexl tests!

**Pull request merged**
